zephyr/drivers/can/can_sccan.c

```diff
@@ -0,0 +1,1032 @@
+	sc_can_write32(config, SCCAN_TMR1_OFFSET, idr);
+
+	/* Write DLC to TMR2 */
+	sc_can_write32(config, SCCAN_TMR2_OFFSET, frame->dlc);
+
+	/* Write CAN Data Frame to TMR3/TMR4 */
+	sc_can_write32(config, SCCAN_TMR3_OFFSET, sys_be32_to_cpu(frame->data_32[0]));
+	sc_can_write32(config, SCCAN_TMR4_OFFSET, sys_be32_to_cpu(frame->data_32[1]));
+
+	/* Save call back function */
+	/* TODO:
+	 * Current SC CAN Controller transmits CAN packets in the order written
+	 * to the TX buffer from user, so call back pointer is stored on the
+	 * premise of FIFO.
+	 * However we plan to improve the sending algorithm on SC CAN controller.
+	 */
+	data->tx_cb[data->tx_head] = callback;
+	data->tx_cb_arg[data->tx_head] = user_data;
+	data->tx_head++;
+	if (data->tx_head == SCCAN_TX_FIFO_DEPTH) {
+		data->tx_head  = 0;
+	}
+
+	k_mutex_unlock(&data->inst_mutex);
+
+	return 0;
+}
+
+/*
+ * SC CAN controller has four "Acceptance filter" feature, so this driver use
+ * it as the "RX filter".
+ * And, due to the specifications of the SC CAN controller, Acceptance Filter
+ * can only be set when disable CAN.
+ * However, Zephyr specification don't defines the timing for adding/removing
+ * RX filter. In fact, CAN sample application (samples/drivers/can/counter) is
+ * called can_add_rx_filter_msgq() after can_start().
+ * Therefore, in this driver, temporarily disabled CAN and set
+ * "Acceptance filer" to SC CAN controller.
+ */
+static int sc_can_add_rx_filter(const struct device *dev, can_rx_callback_t cb,
+				  void *cb_arg, const struct can_filter *filter)
+{
+	const struct sc_can_cfg *config = dev->config;
+	struct sc_can_data *data = dev->data;
+	bool do_disable = false;
+	int ret = 0;
+	int filter_id;
+
+	LOG_DBG("Setting filter ID: 0x%x, mask: 0x%x", filter->id, filter->mask);
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+
+	if (sc_can_is_enabled(config)) {
+		/* Disable CAN with retry if enabled */
+		ret = sc_can_disable(config);
+		if (ret != 0) {
+			goto unlock;
+		}
+		do_disable = true;
+	}
+
+	ret = -ENOSPC;
+	for (filter_id = 0; filter_id < CONFIG_CAN_SCCAN_MAX_FILTER; filter_id++) {
+		if (!sc_can_filter_is_used(config, filter_id)) {
+			sc_can_enable_acceptance_filter(config, filter_id, filter);
+
+			data->rx_cb[filter_id] = cb;
+			data->rx_cb_arg[filter_id] = cb_arg;
+			data->rx_filter[filter_id] = *filter;
+			ret = filter_id;
+			break;
+		}
+	}
+
+	if (filter_id == CONFIG_CAN_SCCAN_MAX_FILTER) {
+		LOG_ERR("No free filter left");
+	} else {
+		LOG_DBG("Filter added. ID: %d", filter_id);
+	}
+
+	if (do_disable) {
+		/* Enable CAN if disabled in this API */
+		sc_can_enable(config);
+	}
+
+unlock:
+	k_mutex_unlock(&data->inst_mutex);
+
+	return ret;
+}
+
+static void sc_can_remove_rx_filter(const struct device *dev, int filter_id)
+{
+	const struct sc_can_cfg *config = dev->config;
+	struct sc_can_data *data = dev->data;
+	int ret;
+	uint32_t enable_reg;
+	bool do_disable = false;
+
+	if (filter_id >= CONFIG_CAN_SCCAN_MAX_FILTER) {
```

Comment:
should `CONFIG_CAN_SCCAN_MAX_FILTER` be in dts?

Reply:
Yes, should be move to dts.
And TX_FIFO_DEPTH too.
Already moved.